### PR TITLE
Improve the bc signer to copy the stream rather than to read all bytes

### PR DESCRIPTION
**Pull request by merks**

### CODE REVIEW

File: tycho-gpg-plugin/src/main/java/org/eclipse/tycho/gpg/BouncyCastleSigner.java [hunk 1]
```diff
@@ -143,7 +144,17 @@ public SignatureStore generateSignature(File file) throws PGPException, IOExcept
         }
         signatureGenerator.setHashedSubpackets(subpackets.generate());
         signatureGenerator.init(PGPSignature.BINARY_DOCUMENT, privateKey);
-        signatureGenerator.update(Files.readAllBytes(file.toPath()));
+        Files.copy(file.toPath(), new OutputStream() {
```
Comment (laeubi) [inline tycho-gpg-plugin/src/main/java/org/eclipse/tycho/gpg/BouncyCastleSigner.java:147]: Maybe a "traditional" loop with a FileInsputstream would be sufficient here?

**Comment by merks** [comment]
I think you are suggesting to inline this method java.io.InputStream.transferTo(OutputStream) to make the code more traditional, but I wonder why that's actually better?
```
    public long transferTo(OutputStream out) throws IOException {
        Objects.requireNonNull(out, "out");
        long transferred = 0;
        byte[] buffer = new byte[DEFAULT_BUFFER_SIZE];
        int read;
        while ((read = this.read(buffer, 0, DEFAULT_BUFFER_SIZE)) >= 0) {
            out.write(buffer, 0, read);
            transferred += read;
        }
        return transferred;
    }
```

**Comment by laeubi** [comment]
I intentionally don't wrote "better" because it is hard to tell what is actually "better" but implementing an Output Stream to read the bytes of an Input stream just feels a bit odd :-)

The "classic" loop would be something like this (if you like buffers):
```
byte[] buffer = new byte[1024*1024*10];
int read;
while ((read = this.read(buffer)) >= 0) {
     signatureGenerator.update(buffer, 0, read);
}
```

**Comment by merks** [comment]
@laeubi The last build aborted so I pushed again to start a new build.

BTW, a test job for this stuff was working earlier:

https://ci.eclipse.org/oomph/job/jetty/job/jetty-mirror-4.0/

But now if fails like this:
```
[INFO] Scanning for projects...
[WARNING] The POM for org.eclipse.platform:org.eclipse.equinox.p2.core:jar:2.9.200-SNAPSHOT is missing, no dependency information available
[WARNING] The POM for org.eclipse.platform:org.eclipse.equinox.registry:jar:3.11.200-SNAPSHOT is missing, no dependency information available
[WARNING] The POM for org.eclipse.platform:org.eclipse.equinox.preferences:jar:3.10.100-SNAPSHOT is missing, no dependency information available
[WARNING] The POM for org.eclipse.platform:org.eclipse.core.contenttype:jar:3.8.200-SNAPSHOT is missing, no dependency information available
[WARNING] The POM for org.eclipse.platform:org.eclipse.equinox.app:jar:1.6.200-SNAPSHOT is missing, no dependency information available
[WARNING] The POM for org.eclipse.platform:org.eclipse.equinox.p2.metadata:jar:2.6.300-SNAPSHOT is missing, no dependency information available
[WARNING] The POM for org.eclipse.platform:org.eclipse.equinox.security:jar:1.3.1000-SNAPSHOT is missing, no dependency information available
[WARNING] The POM for org.eclipse.platform:org.eclipse.equinox.p2.jarprocessor:jar:1.2.300-SNAPSHOT is missing, no dependency information available
[WARNING] The POM for org.eclipse.platform:org.eclipse.equinox.p2.metadata.repository:jar:1.4.100-SNAPSHOT is missing, no dependency information available
[WARNING] The POM for org.eclipse.platform:org.eclipse.equinox.simpleconfigurator:jar:1.4.100-SNAPSHOT is missing, no dependency information available
[WARNING] The POM for org.eclipse.platform:org.eclipse.equinox.p2.engine:jar:2.7.500-SNAPSHOT is missing, no dependency information available
[WARNING] The POM for org.eclipse.platform:org.eclipse.equinox.p2.director:jar:2.5.400-SNAPSHOT is missing, no dependency information available
[WARNING] The POM for org.eclipse.platform:org.eclipse.equinox.frameworkadmin:jar:2.2.100-SNAPSHOT is missing, no dependency information available
[WARNING] The POM for org.eclipse.platform:org.eclipse.equinox.p2.garbagecollector:jar:1.2.0-SNAPSHOT is missing, no dependency information available
[WARNING] The POM for org.eclipse.platform:org.eclipse.equinox.simpleconfigurator.manipulator:jar:2.2.0-SNAPSHOT is missing, no dependency information available
[WARNING] The POM for org.eclipse.platform:org.eclipse.equinox.frameworkadmin.equinox:jar:1.2.200-SNAPSHOT is missing, no dependency information available
[ERROR] [ERROR] Some problems were encountered while processing the POMs:
[ERROR] Unresolveable build extension: Plugin org.eclipse.tycho:tycho-maven-plugin:4.0.0-SNAPSHOT or one of its dependencies could not be resolved: The following artifacts could not be resolved: org.eclipse.platform:org.eclipse.equinox.common:jar:3.17.0-SNAPSHOT, org.eclipse.platform:org.eclipse.core.jobs:jar:3.13.200-SNAPSHOT: Could not find artifact org.eclipse.platform:org.eclipse.equinox.common:jar:3.17.0-SNAPSHOT in eclipse-maven-releases (https://repo.eclipse.org/content/repositories/snapshots) @ 
 @ 
[ERROR] The build could not read 1 project -> [Help 1]
[ERROR]   
[ERROR]   The project org.eclipse.oomph.jetty.mirror:org.eclipse.oomph.jetty.mirror:1.0.0-SNAPSHOT (/home/jenkins/agent/workspace/jetty/jetty-mirror-4.0/jetty/pom.xml) has 1 error
[ERROR]     Unresolveable build extension: Plugin org.eclipse.tycho:tycho-maven-plugin:4.0.0-SNAPSHOT or one of its dependencies could not be resolved: The following artifacts could not be resolved: org.eclipse.platform:org.eclipse.equinox.common:jar:3.17.0-SNAPSHOT, org.eclipse.platform:org.eclipse.core.jobs:jar:3.13.200-SNAPSHOT: Could not find artifact org.eclipse.platform:org.eclipse.equinox.common:jar:3.17.0-SNAPSHOT in eclipse-maven-releases (https://repo.eclipse.org/content/repositories/snapshots) -> [Help 2]
[ERROR]
```
In the previous build there were also warnings, but it didn't fail:
```
[INFO] Scanning for projects...
[WARNING] The POM for org.eclipse.platform:org.eclipse.equinox.p2.core:jar:2.9.200-SNAPSHOT is missing, no dependency information available
[WARNING] The POM for org.eclipse.platform:org.eclipse.equinox.registry:jar:3.11.200-SNAPSHOT is missing, no dependency information available
[WARNING] The POM for org.eclipse.platform:org.eclipse.equinox.preferences:jar:3.10.100-SNAPSHOT is missing, no dependency information available
[WARNING] The POM for org.eclipse.platform:org.eclipse.core.contenttype:jar:3.8.200-SNAPSHOT is missing, no dependency information available
[WARNING] The POM for org.eclipse.platform:org.eclipse.equinox.app:jar:1.6.200-SNAPSHOT is missing, no dependency information available
[WARNING] The POM for org.eclipse.platform:org.eclipse.equinox.p2.metadata:jar:2.6.300-SNAPSHOT is missing, no dependency information available
[WARNING] The POM for org.eclipse.platform:org.eclipse.equinox.security:jar:1.3.1000-SNAPSHOT is missing, no dependency information available
[WARNING] The POM for org.eclipse.platform:org.eclipse.equinox.p2.jarprocessor:jar:1.2.300-SNAPSHOT is missing, no dependency information available
[WARNING] The POM for org.eclipse.platform:org.eclipse.equinox.p2.metadata.repository:jar:1.4.100-SNAPSHOT is missing, no dependency information available
[WARNING] The POM for org.eclipse.platform:org.eclipse.equinox.simpleconfigurator:jar:1.4.100-SNAPSHOT is missing, no dependency information available
[WARNING] The POM for org.eclipse.platform:org.eclipse.equinox.p2.engine:jar:2.7.500-SNAPSHOT is missing, no dependency information available
[WARNING] The POM for org.eclipse.platform:org.eclipse.equinox.p2.director:jar:2.5.400-SNAPSHOT is missing, no dependency information available
[WARNING] The POM for org.eclipse.platform:org.eclipse.equinox.frameworkadmin:jar:2.2.100-SNAPSHOT is missing, no dependency information available
[WARNING] The POM for org.eclipse.platform:org.eclipse.equinox.p2.garbagecollector:jar:1.2.0-SNAPSHOT is missing, no dependency information available
[WARNING] The POM for org.eclipse.platform:org.eclipse.equinox.simpleconfigurator.manipulator:jar:2.2.0-SNAPSHOT is missing, no dependency information available
[WARNING] The POM for org.eclipse.platform:org.eclipse.equinox.frameworkadmin.equinox:jar:1.2.200-SNAPSHOT is missing, no dependency information available
[INFO] Tycho Version:  4.0.0-SNAPSHOT
[INFO] Tycho Resolver: classic
[INFO] Tycho Mode:     project
[INFO] 
```
Any clue what might cause this?  I suppose/hope it will magically just go away again...

**Comment by laeubi** [comment]
> Any clue what might cause this? I suppose/hope it will magically just go away again...

Yeah I don't know what causing this it seems the infra is unstable but I have not yet found a direct error to report to gitlab :-\

**Review by laeubi — APPROVED** [review]
This looks good and the build failures seem unrelated.

**Comment by merks** [comment]
This is happening on my ci.eclipse.org instance and it is accessing https://repo.eclipse.org/content/repositories/snapshots so perhaps the snapshot repo  is sometimes in an inconsistent state...

**Comment by laeubi** [comment]
It seems we have infra-issues (again), I reopened:

- https://gitlab.eclipse.org/eclipsefdn/helpdesk/-/issues/1390

**Comment by merks** [comment]
Thank you yet again!

**Comment by laeubi** [comment]
@merks seems there are still issues after respin, let me know if this is important for you and I can then merge this or we wait until the CI is working again. But we might still not get new snapshots with broken repository server :-[

**Comment by merks** [comment]
There's no urgency.  It's just an optimization...

**Pull request merged**